File: instrumentation/http4s-blaze-server-2.12_0.21/src/main/scala/com/nr/instrumentation/http4s/RequestWrapper.scala
package com.nr.instrumentation.http4s

import cats.effect.Sync
import org.http4s.Request

import java.util
import com.newrelic.api.agent.{ExtendedRequest, HeaderType}
import org.http4s.util.CaseInsensitiveString

import collection.JavaConverters._

class RequestWrapper[F[_] : Sync](request: Request[F]) extends ExtendedRequest {

  def getMethod: String = {
    request.method.name
  }

  def getRequestURI: String = {
    request.uri.path.toString()
  }

  def getRemoteUser: String = {
    null
  }

  def getParameterNames: java.util.Enumeration[_] = {
    request.uri.query.params.keysIterator.asJavaEnumeration
  }

  def getParameterValues(name: String): Array[String] = {
    request.uri.query.multiParams.getOrElse(name, Seq.empty).toArray
  }

  def getAttribute(name: String): AnyRef = {
    null
  }

  def getCookieValue(name: String): String =
    request.cookies
           .find(_.name.equalsIgnoreCase(name))
           .map(_.content)
           .orNull

  def getHeaderType: HeaderType = {
    HeaderType.HTTP
  }

  def getHeader(name: String): String =
    request.headers
           .find(_.name == CaseInsensitiveString(name))
           .map(_.value)
           .orNull

  override def getHeaders(name: String): util.List[String] =
    request.headers
           .filter(_.name == CaseInsensitiveString(name))
           .toList
           .map(_.value)
           .asJava
}

object RequestWrapper {
  def apply[F[_] : Sync](request: Request[F]): ExtendedRequest =
    new RequestWrapper[F](request)
}

File: instrumentation/undertow-server-2.0.0/src/main/java/com/nr/agent/instrumentation/undertow/RequestWrapper.java
package com.nr.agent.instrumentation.undertow;

import com.newrelic.api.agent.ExtendedRequest;
import com.newrelic.api.agent.HeaderType;
import io.undertow.server.HttpServerExchange;
import io.undertow.server.handlers.Cookie;
import io.undertow.util.HeaderMap;
import io.undertow.util.HttpString;

import java.util.Collections;
import java.util.Deque;
import java.util.Enumeration;
import java.util.Map;

public class RequestWrapper extends ExtendedRequest {
    private final Map<String, Cookie> cookies;
    private final Map<String, Deque<String>> params;
    private final String requestUri;
    private final HeaderMap headers;
    private final HttpString method;

    public RequestWrapper(HttpServerExchange e) {
        cookies = e.getRequestCookies();
        params = e.getPathParameters();
        requestUri = e.getRequestURI();
        headers = e.getRequestHeaders();
        method = e.getRequestMethod();
    }

    @Override
    public Object getAttribute(String name) {
        return null;
    }

    @Override
    public String getCookieValue(String name) {
        if (cookies != null) {
            Cookie cookie = cookies.get(name);
            if(cookie != null)
                return cookie.getValue();
        }

        return null;
    }

    @SuppressWarnings("rawtypes")
    @Override
    public Enumeration getParameterNames() {
        if (params != null) {
            return Collections.enumeration(params.keySet());
        }

        return null;
    }

    @Override
    public String[] getParameterValues(String name) {
        if (params != null) {
            Deque<String> values = params.get(name);
            if (values != null) {
                String[] retValue = new String[values.size()];
                values.toArray(retValue);
                return retValue;
            }
        }

        return null;
    }

    @Override
    public String getRemoteUser() {
        return null;
    }

    @Override
    public String getRequestURI() {
        return requestUri;
    }

    @Override
    public String getHeader(String name) {
        if (headers != null) {
            return headers.getFirst(name);
        }

        return null;
    }

    @Override
    public HeaderType getHeaderType() {
        return HeaderType.HTTP;
    }

    @Override
    public String getMethod() {
        return method.toString();
    }
}
